happi/cli.py

```diff
@@ -392,6 +426,389 @@ def transfer(ctx, name: str, target: str):
+        happi.loader.import_class(result.item.device_class)
+        return time.monotonic() - start
+
+    @staticmethod
+    def run_one_benchmark(
+        result: happi.SearchResult,
+        wait_connected: bool,
+    ) -> float:
+        """
+        Create one object and time it.
+        """
+        start = time.monotonic()
+        device = result.get(use_cache=False)
+        if wait_connected:
+            try:
+                device.wait_for_connection(timeout=10.0)
+            except AttributeError:
+                logger.warning(
+                    f'{result["name"]} does not have wait_for_connection.'
+                )
+            except Exception:
+                logger.warning(
+                    'Timeout after 10s while waiting for connection of '
+                    f'{result["name"]}',
+                )
+        return time.monotonic() - start
+
+
+@happi_cli.command()
+@click.pass_context
+@click.option('-d', '--database', 'profile_database', is_flag=True,
+              help='Profile the database loading.')
+@click.option('-i', '--import', 'profile_import', is_flag=True,
+              help='Profile the module importing.')
+@click.option('-o', '--object', 'profile_object', is_flag=True,
+              help='Profile the object instantiation.')
+@click.option('-a', '--all', 'profile_all', is_flag=True,
+              help='Shortcut for enabling all profile stages.')
+@click.option('-p', '--profiler', default='auto',
+              help='Select which profiler to use.')
+@click.option('--glob/--regex', 'use_glob', default=True,
+              help='Use glob style (default) or regex style search terms. '
+              r'Regex requires backslashes to be escaped (eg. at\\d.\\d)')
+@click.argument('search_criteria', nargs=-1)
+def profile(
+    ctx,
+    profile_database: bool,
+    profile_import: bool,
+    profile_object: bool,
+    profile_all: bool,
+    profiler: str,
+    use_glob: bool,
+    search_criteria: List[str],
+):
+    """
+    Per-function startup speed diagnostic.
+
+    This will go through the happi loading process and show
+    information about the execution time of all the
+    functions called during the process.
+
+    Contains options for picking which devices to check and which
+    part of the loading process to profile. You can choose to
+    profile the happi database loading (-d, --database), the
+    class imports (-i, --import), the object instantiation
+    (-o, --object), or all of the above (-a, --all).
+
+    By default this will use whichever profiler you have installed,
+    but this can also be overriden with the (-p, --profiler) option.
+    The priority order is, first, the pcdsutils line_profiler wrapper
+    (--profiler pcdsutils), and second, the built-in cProfile module
+    (--profiler cprofile). More options may be added later.
+
+    Search terms are standard as in the same search terms as the search
+    cli function. A blank search term means to load all the devices.
```

Comment:
I tried this once and segfaulted.  I need to look into this more to give a better bug report, but I got a lot of connection timeouts